assets/javascripts/discourse/pre-initializers/rich-editor.extension.js

```diff
@@ -0,0 +1,136 @@
+import { withPluginApi } from "discourse/lib/plugin-api";
+import I18n, { i18n } from "discourse-i18n";
+import guessDateFormat from "discourse/plugins/discourse-calendar/discourse/lib/guess-best-date-format";
+
+const EVENT_ATTRIBUTES = {
+  name: { default: null },
+  start: { default: null },
+  end: { default: null },
+  reminders: { default: null },
+  minimal: { default: null },
+  closed: { default: null },
+  status: { default: "public" },
+  timezone: { default: "UTC" },
+  allowedGroups: { default: null },
+};
+
+/** @type {RichEditorExtension} */
+const extension = {
+  nodeSpec: {
+    event: {
+      attrs: EVENT_ATTRIBUTES,
+      group: "block",
+      content: "block+",
+      defining: true,
+      isolating: true,
+      draggable: true,
+      parseDOM: [
+        {
+          tag: "div.discourse-post-event",
+          getAttrs(dom) {
+            return { ...dom.dataset };
+          },
+        },
+      ],
+      toDOM(node) {
+        const dataAttrs = Object.entries(node.attrs).reduce(
+          (acc, [key, value]) => {
+            if (value !== null) {
+              acc[`data-${key.replace(/([A-Z])/g, "-$1").toLowerCase()}`] =
+                value;
+            }
+            return acc;
+          },
+          {}
+        );
+
+        const attrs = {
+          class: "discourse-post-event-preview",
+          ...dataAttrs,
+        };
+
+        const domSpec = ["div", attrs];
+
+        const statusLocaleKey = `discourse_post_event.models.event.status.${node.attrs.status}.title`;
+        if (I18n.lookup(statusLocaleKey, { locale: "en" })) {
+          domSpec.push([
+            "div",
+            { class: "event-preview-status" },
+            i18n(statusLocaleKey),
+          ]);
+        }
+
+        const startsAt = moment.tz(node.attrs.start, node.attrs.timezone);
+        const endsAt =
+          node.attrs.end && moment.tz(node.attrs.end, node.attrs.timezone);
+        const format = guessDateFormat(startsAt, endsAt);
+        const formattedStartsAt = startsAt.tz(moment.tz.guess()).format(format);
+        const datesElement = ["div", { class: "event-preview-dates" }];
+        datesElement.push(["span", { class: "start" }, formattedStartsAt]);
+        domSpec.push(datesElement);
+
+        return domSpec;
+      },
+    },
+  },
+
+  parse: {
+    wrap_bbcode(state, token) {
+      if (token.tag === "div") {
+        if (token.nesting === -1) {
+          state.closeNode();
+          return true;
+        }
+
+        if (token.nesting === 1) {
```

Comment:
this feels too broad, maybe...
```suggestion
      if (token.tag === "div") {
        if (token.nesting === -1 && state.top().type.name === "event") {
          state.closeNode();
          return true;
        }

        if (token.nesting === 1 && token.attrGet("class") === "discourse-post-event") {
```
this also makes me wonder if we couldn't use a different token on markdown-it, we're opening/closing this token but we don't expect any content inside it, do we?

Reply:
Ah yes I was trying to do this sooner but forgot, thanks